### PR TITLE
Add migration code for installed packages

### PR DESCRIPTION
**Pull request by CalumMcCall**
- PRO-2111
 - Add method to move data on software installed on devices from
   resolver to device registry.
 - Had to add additional endpoint to work around deadlock bug in slick.

### CODE REVIEW

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        case row@InstalledForeignPackage(uuid, _, _) =>
+          deviceRegistry.fetchMyDevice(uuid)
+            .map { d =>
+              None
+            }
+            .recover {
+              case ex => Option(row)
+            }
+      }
+      Future.sequence(registryDevicesOps).map(_.collect {
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: this collect can be replaced with `_.flatten`

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  def apply(deviceRegistry: DeviceRegistry)(implicit db: Database, ec: ExecutionContext): Unit = {
+
+    val resultF = db.run(ForeignPackages.foreignPackages.result).map { rows =>
+      val registryDevicesOps: Seq[Future[Option[InstalledForeignPackage]]] = rows.map {
+        case row@InstalledForeignPackage(uuid, _, _) =>
+          deviceRegistry.fetchMyDevice(uuid)
+            .map { d =>
+              None
+            }
+            .recover {
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: we should probably check what is the exception here.

File: external-resolver/src/main/scala/org/genivi/sota/resolver/db/ForeignPackages.scala [hunk 3]
```diff
@@ -44,7 +44,7 @@ object ForeignPackages {
     def * = (device, name, version, lastModified) <> (fromTuple, toTuple)
   }
 
-  private val foreignPackages = TableQuery[InstalledForeignPackageTable]
+  val foreignPackages = TableQuery[InstalledForeignPackageTable]
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/db/ForeignPackages.scala]: we should not give access to this field. It was `private` for a reason...

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala [hunk 4]
```diff
@@ -140,15 +148,20 @@ class DevicesResource(namespaceExtractor: Directive1[AuthedNamespaceScope],
...
-        fetchDevice(uuid)
+    pathPrefix("mydevice") {
+      extractUuid { uuid =>
+        (post & path("ping") & authedNs.oauthScope(s"ota-core.${uuid.show}.write")) {
+          updateLastSeen(uuid)
+        } ~
+        (get & pathEnd & authedNs.oauthScopeReadonly(s"ota-core.${uuid.show}.read")) {
+          fetchDevice(uuid)
+        }
+        (put & path("packages") & authedNs.oauthScope(s"ota-core.{device.show}.write")) {
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: why was this added?

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala [hunk 5]
```diff
@@ -102,6 +103,13 @@ class DevicesResource(namespaceExtractor: Directive1[AuthedNamespaceScope],
     }
   }
 
+  def resolverPackageMigration(): Route = {
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: will this be removed afterwards?

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+object DeviceDataMigrator {
+
+  private lazy val logger = LoggerFactory.getLogger(this.getClass)
+
+  def apply(deviceRegistry: DeviceRegistry)(implicit db: Database, ec: ExecutionContext): Unit = {
+
+    val resultF = db.run(ForeignPackages.foreignPackages.result).map { rows =>
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: if this works, for now it's fine.

In the future you might want to look at `db.stream`. Then you can add a `Sink` on that stream to process each row individually and contact device registry. This way you get automagic backpressure support and error handling, without even needing to put the whole table in memory. You also won't need to wait on that future since it will just run until done.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala (hunk 4 above)
Comment (Danten) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: `s"ota-core.{device.show}.write"` should be `s"ota-core.${uuid.show}.write"`

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala [hunk 7]
```diff
@@ -53,6 +53,9 @@ object InstalledPackages {
       installedPackages ++= packages.map(InstalledPackage(device, _, Instant.now()))
     ).transactionally
 
+  def addRows(rows: Seq[InstalledPackage])(implicit ec: ExecutionContext): DBIO[Seq[Int]] =
+    DBIO.sequence(rows.map(installedPackages += _)).transactionally
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala]: This whole migration should be idempotent, I think this prevents the migration from being idempotent, it will just add more and more rows, or will just fail because we are trying to add duplicate keys.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala [hunk 8]
```diff
@@ -140,15 +148,20 @@ class DevicesResource(namespaceExtractor: Directive1[AuthedNamespaceScope],
...
+        (get & pathEnd & authedNs.oauthScopeReadonly(s"ota-core.${uuid.show}.read")) {
+          fetchDevice(uuid)
+        }
+        (put & path("packages") & authedNs.oauthScope(s"ota-core.{device.show}.write")) {
+           updateInstalledSoftware(uuid)
+        }
       } ~
       (put & path("packages") & authedNs.oauthScope(s"ota-core.{device.show}.write")) {
-        updateInstalledSoftware(uuid)
+        resolverPackageMigration()
```
Comment (Danten) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: This route doesn't seem to originate from a device, so it should probably not be in mydevice. Also the scope is wrong.

File: common-client/src/main/scala/org/genivi/sota/DeviceRegistry.scala [hunk 9]
```diff
@@ -53,4 +55,7 @@ trait DeviceRegistry {
 
   def setInstalledPackages
     (device: Uuid, packages: Seq[PackageId])(implicit ec: ExecutionContext) : Future[NoContent]
+
+  def setInstalledPackagesRows
+    (rows: Json)(implicit ec: ExecutionContext) : Future[NoContent]
```
Comment (Danten) [inline common-client/src/main/scala/org/genivi/sota/DeviceRegistry.scala]: Why using `Json` instead of `Seq[InstalledPackages]`?

File: external-resolver/src/main/scala/org/genivi/sota/resolver/db/ForeignPackages.scala (hunk 3 above)
Comment (CalumMcCall) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/db/ForeignPackages.scala]: I was planning to remove it afterwards, it saves having to create a new temporary method. I don't mind adding a method if you prefer though.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala (hunk 7 above)
Comment (CalumMcCall) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala]: Yeah, I envisioned it either adding rows or failing due to duplicate keys. How is that not idempotent? Sure, it might error if we run it twice, but that wouldn't be disastrous.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala (hunk 8 above)
Comment (CalumMcCall) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/DevicesResource.scala]: It doesn't come from a device, true. However, InstalledPackages on device registry has no namespace field, so a namespace for this data doesn't really make sense. Putting it in mydevices was an attempt to bypass authentication as much as possible. I'm open to ideas though.

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala (hunk 7 above)
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/InstalledPackages.scala]: It's not disastrous, but the migration will not run the second time.

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+          case Failure(ex) =>
+            logger.error(s"Failed to write ${packagesPerDevice.toString()} to resolver: ${ex.toString}")
+          case Success(_) =>
+        }
+      case ex =>
+        logger.error(s"failed for reason ${ex.toString}")
+    }
+  }
+
+  def packages()(implicit db: Database, ec: ExecutionContext): DBIO[Seq[InstalledForeignPackage]] =
```
Comment (Danten) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: Let this method return a `Set` to remove one conversion

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    for {
+      nonForeignPkgRows <- DeviceRepository.listInstalledPackages
+      //use sets here to remove duplicates
+      nonForeignPkgs = nonForeignPkgRows.map{r => InstalledForeignPackage(r._2, r._3, Instant.now())}.toSet
```
Comment (Danten) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: Need to use a constant time can't use `Instant.now()` since it will be different for each element in the set (and therefore not remove duplicates of device, packageId).

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+object DeviceDataMigrator {
+
+  private lazy val logger = LoggerFactory.getLogger(this.getClass)
+
+  def apply(deviceRegistry: DeviceRegistry)(implicit db: Database, ec: ExecutionContext): Unit = {
+    db.run(packages()).onComplete {
+      case Success(rows) =>
+        val packagesPerDevice = mutable.Map[Uuid, Seq[PackageId]]()
+        rows.foreach { elem =>
+          packagesPerDevice(elem.device) = packagesPerDevice.getOrElse(elem.device, List()) :+ elem.packageId
```
Comment (Danten) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: You want `+:` here

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success}
+
+object DeviceDataMigrator {
+
+  private lazy val logger = LoggerFactory.getLogger(this.getClass)
+
+  def apply(deviceRegistry: DeviceRegistry)(implicit db: Database, ec: ExecutionContext): Unit = {
+    db.run(packages()).onComplete {
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: Using `onComplete` will install a callback on the future but we never know if it actually runs or runs partially, if the jvm quits before the callback finishes, for example, the migration could run partially. 

`apply` should either return a future, and then boot uses `Await` to wait for the result of the future, or we could use `Await` here, but we should not let this callback run orphan without any checking if it actually succeeds.

File: external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+import scala.concurrent.duration._
+
+object DeviceDataMigrator {
+
+  private lazy val logger = LoggerFactory.getLogger(this.getClass)
+
+  def apply(deviceRegistry: DeviceRegistry)(implicit db: Database, ec: ExecutionContext): Unit = {
+    val f = for {
+      rows          <- db.run(packages())
+      pkgsPerDevice = mutable.Map[Uuid, Set[PackageId]]()
```
Comment (simao) [inline external-resolver/src/main/scala/org/genivi/sota/resolver/DeviceDataMigrator.scala]: ok for code we will delete later, but let's not do this too much ;)
